cmd/kube-apiserver/app/server.go

```diff
@@ -641,7 +643,33 @@ func Complete(s *options.ServerRunOptions) (completedServerRunOptions, error) {
 	return options, nil
 }
 
+var (
+	testServiceResolverLock sync.Mutex
+	testServiceResolver     webhook.ServiceResolver
+)
+
+func SetServiceResolverForTests(t *testing.T, resolver webhook.ServiceResolver) {
```

Comment:
you can document this as not being safe to call concurrently with server start, skip the locks here, and return a cleanup function test code can pass to `t.Cleanup`

Reply:
I'd also maybe consider panicing here if something calls this and testServiceResolver is non-nil... that means a previous test didn't clean up correctly or tests are running concurrently... either way isn't going to be a great situation